### ISSUE TITLE
false positives of cascade_invocations when there are implicit downcast assignments

### ISSUE THREAD

**Issue by yyoon** [dart-lang/linter#787]
Here's an example snippet.

```dart
class Foo {}

class Bar extends Foo {
  void bar() {}
}

void baz(Foo foo) {
  Bar bar = foo;
  bar.bar();      // This line is reported by cascade_invocations rule
}
```

In this case, we can't simply change this to `Bar bar = foo..bar();` because `foo` is not of type `Bar` until it is assigned to `bar` variable using implicit downcast. The only reasonable way I found to get around this issue is to suppress the warning with an `// ignore: cascade_invocations` comment, but maybe I missed something.

**Comment by a14n** [comment]
an other workaround `(foo as Bar).bar()`

**Comment by yyoon** [comment]
Yes, that works. Although we have `avoid_as` turned on as well, so it's not an option for us.

**Comment by zoechi** [comment]
Is there still a good reason for "avoid as"?
I thought that was for performance reasons that are no longer relevant.

**Comment by asashour** [comment]
The case is not compilable, for line

    Bar bar = foo;


>A value of type 'Foo' can't be assigned to a variable of type 'Bar'. (Documentation)
>Try changing the type of the variable, or casting the right-hand type to 'Bar'.

**Comment by bwilkerson** [comment]
The original example was written prior to null safety. After null safely there would need to be either an explicit cast or a type test.

I suspect that we can close this issue.

**Comment by pq** [comment]
I agree.  I'm guessing there's a host of issues like this.
